### PR TITLE
[Flight] Transfer Debug Info in Server-to-Server Flight Requests

### PR DESCRIPTION
**Pull request by sebmarkbage**
A Flight Server can be a consumer of a stream from another Server. In this case the meta data is attached to debugInfo properties on lazy, Promises, Arrays or Elements that might in turn get forwarded to the next stream. In this case we want to forward this debug information to the client in the stream.

I also added a DEV only `environmentName` option to the Flight Server. This lets you name the server that is producing the debug info so that you can trace the origin of where that component is executing. This defaults to `"server"`. DevTools could use this for badges or different colors.

### CODE REVIEW

File: packages/react-client/src/__tests__/ReactFlight-test.js [hunk 1]
```diff
@@ -1806,4 +1806,68 @@ describe('ReactFlight', () => {
...
+      default: <ThirdPartyLazyComponent />,
+    }));
+
+    function ThirdPartyComponent() {
+      return <span>stranger</span>;
+    }
+
+    function ServerComponent({transport}) {
+      // This is a Server Component that receives other Server Components from a third party.
+      const children = ReactNoopFlightClient.read(transport);
```
[the remark this answers is not in the document]
Comment (lubieowoce) [reply packages/react-client/src/__tests__/ReactFlight-test.js:1825]: (for context, the usecase is federation via RSC, as in https://x.com/dan_abramov2/status/1747983201748861274 for example. currently it requires... some stream serialization acrobatics)

[the remark this answers is not in the document]
Comment (sebmarkbage) [reply packages/react-client/src/__tests__/ReactFlight-test.js:1825]: Like anything else it's more of a bundler feature. You just need some way from the third party server to refer to "client" as meaning files in the "server". Like anything else the server needs a manifest from the client.

[the remark this answers is not in the document]
Comment (sebmarkbage) [reply packages/react-client/src/__tests__/ReactFlight-test.js:1825]: Btw, untrusted is NOT supported. Flight parsing isn't vetted from a security perspective if the protocol is from an untrusted source.

[the remark this answers is not in the document]
Comment (lubieowoce) [reply packages/react-client/src/__tests__/ReactFlight-test.js:1825]: i think what @unstubbable was talking about is something that'd allow incorporating foreign (but likely trusted) trees server-side, without going into client-land. right now this seems to require tunneling one RSC stream over another and then deserializing in client-land, as seen here:
https://twitter.com/lubieowoce/status/1744854538060771614?t=IK4O45iZvCU4isBmTHIf1A
and here:
https://twitter.com/ebey_jacob/status/1744793367085727952?t=fOPDaYf3LLelv7gYhvSuzQ&s=19
which works but seems unnecessarily cumbersome i think

basically, i think the feature request would be for a server-land `createFromReadableStream` that keeps client references as references, so that it can be returned from a server component like any other element. (it's unfortunate that it'd have to parse just to serialize it right back, but i'm not sure if that can be avoided -- at the very least, rows need renumbering)

ofc the federated use-case ALSO needs bundler support, because eventually you need to resolve those foreign client references to something. but i believe that's been implemented in userspace already

[the remark this answers is not in the document]
Comment (sebmarkbage) [reply packages/react-client/src/__tests__/ReactFlight-test.js:1825]: Yea, I mean we'll support ReadableStream pass through but that's not really recommended. That's just module federation.

It needs to be deserialized and re-serialized so that the third-party can also render server components and everything else can kick in. But so does SSR. There could be optimizations added on top of that which can keep some parts pass through but that's an optimization and not the core model.

[the remark this answers is not in the document]
Comment (sebmarkbage) [reply packages/react-client/src/__tests__/ReactFlight-test.js:1825]: > basically, i think the feature request would be for a server-land createFromReadableStream that keeps client references as references, so that it can be returned from a server component like any other element. (it's unfortunate that it'd have to parse just to serialize it right back, but i'm not sure if that can be avoided

Although now I'm confused what you mean because this is supported. That's what this PR uses.

There's no limitation that `/client` can't be used in `react-server` environments.

https://github.com/facebook/react/blob/main/packages/react-server-dom-webpack/package.json#L46-L49

In fact, this is how encryption of closures work in Next.js.

[the remark this answers is not in the document]
Comment (sebmarkbage) [reply packages/react-client/src/__tests__/ReactFlight-test.js:1825]: Oh I think I understand the misunderstanding. The confusion is that Client References isn't actually 1:1 with "use client". It's not supposed to preserve the Client References. Instead, the Client Reference of a Third Party is a reference to a module in the first party. Requiring that module might be a Server Component but if it's "use client" then importing that same file will be a Client Reference again. So you don't need some way to do the rewriting.

File: packages/react-server/src/ReactFlightServer.js [hunk 2]
```diff
@@ -1218,9 +1275,30 @@ function renderModelDestructive(
         );
       }
       case REACT_LAZY_TYPE: {
-        const payload = (value: any)._payload;
-        const init = (value: any)._init;
+        // Reset the task's thenable state before continuing. If there was one, it was
+        // from suspending the lazy before.
+        task.thenableState = null;
```
Comment (hoxyq) [inline packages/react-server/src/ReactFlightServer.js:1280]: Maybe a silly question, but why adding / changing `debugInfo` has side-effects on `task.thenableState`?
Or is it just a small drive-by fix?

**Review by hoxyq — COMMENTED** [review]
LGTM from the approach side. Maybe worth req

File: packages/react-server/src/ReactFlightServer.js [hunk 3]
```diff
@@ -199,6 +202,8 @@ export type Request = {
   taintCleanupQueue: Array<string | bigint>,
   onError: (error: mixed) => ?string,
   onPostpone: (reason: string) => void,
+  // DEV-only
+  environmentName: string,
```
[the remark this answers is not in the document]
Comment (gnoff) [reply packages/react-server/src/ReactFlightServer.js:206]: Yeah personally I'd prefer we any cast the one place we read the name than do this to the Request type

File: packages/react-server/src/ReactFlightServer.js [hunk 4]
```diff
@@ -1218,9 +1275,30 @@ function renderModelDestructive(
...
+        const init = lazy._init;
         const resolvedModel = init(payload);
+        if (__DEV__) {
+          const debugInfo: ?ReactDebugInfo = lazy._debugInfo;
+          if (debugInfo) {
+            // If this came from Flight, forward any debug info into this new row.
+            if (debugID === null) {
+              // We don't have a chunk to assign debug info. We need to outline this
+              // component to assign it an ID.
+              return outlineTask(request, task);
```
Comment (gnoff) [inline packages/react-server/src/ReactFlightServer.js:1293]: It seems unfortunate that we fork the dev/prod behavior here. Means we need to have faith that outlining won't hide or express bugs that won't happen absent the debugInfo.

**Review by gnoff — APPROVED** [review]

File: packages/react-server/src/ReactFlightServer.js (hunk 4 above)
Comment (sebmarkbage) [reply packages/react-server/src/ReactFlightServer.js:1293]: Well it would show up in prod. As long as it's unobservable it doesn't affect dev/prod behavior for users. At some point we just have to trust that we don't mess up the implementation.

File: packages/react-server/src/ReactFlightServer.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (sebmarkbage) [reply packages/react-server/src/ReactFlightServer.js:206]: We have had this pattern elsewhere. It's nice because there's only one place to mess it up.

There will be many more places where `environmentName` gets used but the Request type is only initialized in one place.

Optional fields are bad for perf when you do mess them up because it breaks the hidden class. You could potentially accidentally add a field by writing to it unconditionally.

I don't like that we use optional fields for Fiber for example because we instead have bugs in the continued working with the code by doing undefined checks, which lets things pass and keep working even if they're undefined or missing. We typically check for example `=== null` exactly so that if we get an undefined flowing through the system it should error. Even if that errors only in production because then we'll quickly find the mistake.

File: packages/react-server/src/ReactFlightServer.js (hunk 2 above)
Comment (sebmarkbage) [reply packages/react-server/src/ReactFlightServer.js:1280]: Yea, it's mainly a drive-by fix. The bug was introduced in https://github.com/facebook/react/pull/28068
